### ISSUE TITLE
Problem with multiple entries from the _same_ cached tar archive

### ISSUE THREAD

**Issue by willirath** [intake/intake#499]
With two 

```shell
$ rm -rf ~/.intake/
$ mkdir -p two_csvs/
$ seq 0 9 | paste -sd, > two_csvs/data_01.csv
$ seq 20 40 | paste -sd, > two_csvs/data_02.csv
$ tar -czf two_csvs.tar.gz two_csvs/
$ cat two_csvs/data_01.csv
0,1,2,3,4,5,6,7,8,9
$ cat two_csvs/data_02.csv
20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40
```

`catalog.yaml`:

```yaml
metadata:
  version: 1

sources:
  data_01:
    driver: csv
    args:
      urlpath: "two_csvs.tar.gz"
    cache:
      - type: compressed
        decomp: tgz
        argkey: urlpath
        regex_filter: 'data_01.csv'
            
  data_02:
    driver: csv
    args:
      urlpath: "two_csvs.tar.gz"
    cache:
      - type: compressed
        decomp: tgz
        argkey: urlpath
        regex_filter: 'data_02.csv'
```

```python
import intake
cat = intake.open_catalog("catalog.yaml")
list(cat)
```
```
['data_01', 'data_02']
```

```python
print(list(cat["data_01"].read()))  # truth: 0..9
```
```
['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
```

```python
print(list(cat["data_02"].read()))  # truth: 20..40
```
```
['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
```

```python
cat["data_02"].cache[0].clear_all()
print(list(cat["data_02"].read()))  # truth: 20..40
```
```
['20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40']
```

```python
print(list(cat["data_01"].read()))  # truth: 0..9
```
```
['20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40']
```

**Comment by martindurant** [comment]
Yes, it appears that the urlpath is being hashed, and of course evaluates as the same for both cases, regardless of the target file - which sort of makes sense, but it not useful for you.

The "cache:" blocks in Intake are complex to implement and use, so there has been an effort to transfer the functionality to [fsspec](https://filesystem-spec.readthedocs.io/en/latest/features.html#caching-files-locally). This works well with, for instance, zip-files, where instead of the "cache:" block, you might replace urlpath with something like "simplecache::zip://data_01.csv::file://myzip.zip". OK, so maybe that doesn't look *that* simple, but you can read it (with "::" as the delimiter, right-to-left) as
- read from the file myzip.zip
- treat this as a zip filesystem, and target file data_01.csv
- cache this on first use (the cache location is configurable)
Unfortunately, there is no driver for tar because, unlike zip, it doesn't contain an index. But it might [still be reasonable](https://github.com/intake/filesystem_spec/issues/297) (and easy) to write that implementation, especially for uncompressed tar.

**Comment by willirath** [comment]
Thanks @martindurant, this is very helpful! In my use case, I can easily control / influence the choice of the archive format and will just switch to zip.

I share the impression that having a tar / compressed-tar driver would make sense even if it comes with a huge performance penalty. 

Should I close here? Or do you want this kept open?

**Comment by martindurant** [comment]
I don't mind whether it is left open - it wouldn't be "resolved" until there is indeed a tar driver in fsspec.

**Comment by willirath** [comment]
With 
```yaml
metadata:
  version: 1

sources:
  data_01:
    driver: csv
    args:
      urlpath: "simplecache::zip://data_01.csv::file://two_csvs.zip"
```
I get
```python-traceback
[...]
/srv/conda/envs/notebook/lib/python3.7/site-packages/fsspec/registry.py in get_filesystem_class(protocol)
    177     if protocol not in registry:
    178         if protocol not in known_implementations:
--> 179             raise ValueError("Protocol not known: %s" % protocol)
    180         bit = known_implementations[protocol]
    181         try:

ValueError: Protocol not known: simplecache::zip
```
with `work/` being the current directory for the Python process.

Dropping the `simplecache` (`urlpath: "zip://data_01.csv::file://two_csvs.zip"`) leads to:
```python-traceback
[...]
/srv/conda/envs/notebook/lib/python3.7/site-packages/fsspec/implementations/local.py in _open(self)
    190         if self.f is None or self.f.closed:
    191             if self.autocommit or "w" not in self.mode:
--> 192                 self.f = open(self.path, mode=self.mode)
    193             else:
    194                 # TODO: check if path is writable?

IsADirectoryError: [Errno 21] Is a directory: '/work'
```

Curiously, if I directly use `fsspec` to open `simplecache::zip://data_01.csv::file://two_csvs.zip` and pass it to `pandas.read_csv` everything looks fine.

**Comment by willirath** [comment]
This is the case for the current pypi releases of intake and fsspec and also for intake and fsspec from master.

**Comment by martindurant** [comment]
Should be fixed by https://github.com/intake/filesystem_spec/pull/307

**Comment by willirath** [comment]
> Should be fixed by intake/filesystem_spec#307

Confirmed. Awesome!
